src/edu/washington/escience/myria/expression/evaluate/GenericEvaluator.java

```diff
@@ -108,14 +129,36 @@ public void compile() throws DbException {
    * @param inputRow index of the row that should be used for input data
    * @param state additional state that affects the computation
    * @param stateRow index of the row that should be used for state
+   * @param stateColOffset the column offset of the state
+   * @throws DbException in case of error.
+   */
+  public void updateState(
+      @Nonnull final ReadableTable input,
+      final int inputRow,
+      @Nonnull final MutableTupleBuffer state,
+      final int stateRow,
+      final int stateColOffset)
+      throws DbException {
+    ((ExpressionEvalReplaceInterface) evaluator)
+        .evaluate(input, inputRow, state, stateRow, stateColOffset);
+  }
+
+  /**
+   * Evaluates the {@link #getJavaExpressionWithAppend()} using the {@link #evaluator}. Prefer to use
+   * {@link #evaluateColumn(TupleBatch)} since it can evaluate an entire TupleBatch at a time for better locality.
+   *
+   * @param input a tuple batch
+   * @param inputRow index of the row that should be used for input data
+   * @param state additional state that affects the computation
+   * @param stateRow index of the row that should be used for state
    * @param result the table storing the result
    * @param count column storing number of results (null for single-valued expressions)
    * @throws DbException in case of error.
    */
   public void eval(
-      @Nonnull final ReadableTable input,
+      @Nullable final ReadableTable input,
```

Comment:
When an emit expression is evaluated, the `input` is null but the `state` is not, although the `state` works as an `input`.